### ISSUE TITLE
Provide the ability to request complete add-ons via the Search API

### ISSUE THREAD

**Issue by bobsilverberg** [mozilla/addons-server#10232]
This is being opened to address https://github.com/mozilla/addons-frontend/issues/7134. We are having an issue because the [Search API](https://addons-server.readthedocs.io/en/latest/topics/api/addons.html#search) returns "partial" add-ons (it excludes all `license` fields and `release_notes`), but we use it for the guides pages which then load the add-ons into the redux state. When those add-ons are encountered on an add-on detail page, they are not re-fetched because they already exist in state, and therefore the necessary `license` and `release_notes` data is missing.

We could like to be able to pass a flag to the Search API to request that "complete" add-ons be returned. These would include `license` fields (although not `license.text`) and `release_notes`. Essentially the add-on data returned by the Search API, when "complete" add-ons are requested, would be the same as that returned by the [Detail API](https://addons-server.readthedocs.io/en/latest/topics/api/addons.html#detail).

The default behaviour of the Search API would be to return "partial" add-on data, just as it works now, and the only time we currently anticipate asking for "complete" add-on data is when requesting add-ons for the guides pages.

**Comment by diox** [comment]
I can't find an appropriate name for the parameter to return that extra data... and so I'm tempted to *always* return `license` (without the `license.text`) and `release_notes`. This would have a slight performance penalty, but would make things much simpler. I'm not a fan of how our pseudo-REST API is looking more and more like a poor man's GraphQL API :)

Thoughts @eviljeff @EnTeQuAk (and frontend people as well) ?

**Comment by EnTeQuAk** [comment]
Personally, we should make sure to use gzip in our API at some point (or are we already?) and just return the extra data if it's needed most of the time anyway (not sure here, @bobsilverberg can enlighten us where exactly the additional data is requested to get an idea). That way the data can be cached more easily and the performance penalty shouldn't be too bad. That'd be a very naive approach.

A different approach would be to indeed add a field `force_full_addon_data=True` or so and also have a field `partial=True/False` that indicates what version we actually render (may help debug this at some point).

**Comment by diox** [comment]
We're already gzipping the API, yes.

The "full" version would only be needed in the guides pages for now. 

The problem I have with the query parameter approach is that it's not really a full/complete add-on, as we would still be missing the license text as well as the picture url for authors, and implementing those is a lot more annoying/costly.

**Comment by EnTeQuAk** [comment]
> The problem I have with the query parameter approach is that it's not really a full/complete add-on, as we would still be missing the license text as well as the picture url for authors, and implementing those is a lot more annoying/costly.

Then, maybe go the other way around and add a `extra_fields` parameter that works similar to the drf supported `fields` but only adds additional fields to the serialization? That's, even more, a graphql-ish way of doing things but at least supports what we need.

**Comment by eviljeff** [comment]
> I'm not a fan of how our pseudo-REST API is looking more and more like a poor man's GraphQL API

Yep.  That's my concern.  We've already got too many flags really and adding more complexity doesn't help the codebase maintenance or API consumers in the end.

Is this genuinely a case when we need `release notes` and some license details for _search_ type response where multiple results are displayed (I wasn't clear from the linked issue)?  If so let's either add them for all results or create a new endpoint just for that use case.
If not then, IMO, frontend should make another request for the detail response to get all the fields.

**Comment by diox** [comment]
The guides pages use the search API with multiple results (using the `guid` parameter + multiple guids) but don't *directly* use the release notes and license details - it stores the results as add-on objects and *then* when visiting the detail page the frontend code assumes that everything was fetched when in fact it's missing those 2 things.

**Comment by eviljeff** [comment]
> then when visiting the detail page the frontend code assumes that everything was fetched

That's the assumption that should be changed, imo.

But, meh, just exposing release notes and license details I suppose works too (though release notes could be huuuuge, afaik)

**Comment by diox** [comment]
> @diox when you say all license info (except full text) do you mean the text added in dev hub to addons with a Custom license?

Yes.
